File: golang/architecture-shell-cqrs/base_request.go
package cqrs

// BaseRequest is the common ancestor for all application-layer requests.
// Enables uniform pipeline processing for commands, queries, and future request types.
// All commands, queries, notifications, and jobs must implement this interface.
type BaseRequest interface {
	// IsRequest is a marker method to identify base requests.
	// This method provides type-safe mediator registration and resolution.
	IsRequest()
}

File: typescript-nodejs/architecture-shell-cqrs/src/BaseRequest.ts
/**
 * BaseRequest - Base type for all application-layer requests
 *
 * Common ancestor for commands, queries, and other request types,
 * enabling uniform pipeline processing.
 *
 * @typeParam TResponse - The response type (defaults to void)
 *
 * @remarks
 * All requests (commands, queries, notifications, jobs) must extend this type
 * to enable type-safe mediator registration and resolution.
 *
 * @see {@link Command} - State-changing operations
 * @see {@link Query} - Read-only operations
 */
export interface BaseRequest<TResponse = void> {
  // Marker interface - TResponse is a phantom type for compile-time type safety
  __phantom?: TResponse;
}
